rapid/src/main/java/com/vrg/rapid/MembershipService.java

```diff
@@ -494,9 +509,36 @@ int getMembershipSize() {
      * Shuts down all the executors.
      */
     void shutdown() {
+        leave();
         alertBatcherJob.cancel(true);
         failureDetectorJobs.forEach(k -> k.cancel(true));
         messagingClient.shutdown();
+        membershipView.reset();
```

Comment:
Regarding the view reset... are you concerned that a user might try to read the membership view after shutdown() is called?

Reply:
Yes - in fact that's what was happening in the tests. I realized after implementing it that it was much simpler to kick out the leaving node of the test scaffold but left this here "just in case". I mean I could see someone erroneously calling the service after the node left or was kicked (if they don't handle the event properly). I'm not sure as to whether it's necessary, just seemed like that way we'd have a consistent view exposed at all times

Reply:
I think it's better to throw an exception if any of the APIs are invoked after shutdown, than to return a cleared out membership view.

Also, in the tests where we check for failures, putting leave() inside shutdown(){} would make all those tests use a graceful leave instead of failing like before. It might be better to add a leave() API to Cluster.class (and use the hard shutdown option only for tests).